src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -90,30 +94,83 @@ async Task AuthenticateSASL(List<string> mechanisms, bool async)
             if (!firstServerMsg.Nonce.StartsWith(clientNonce))
                 throw new InvalidOperationException("[SCRAM] Malformed SCRAMServerFirst message: server nonce doesn't start with client nonce");
 
-            var scramFinalClientMsg = new SCRAMClientFinalMessage(passwd, firstServerMsg.Nonce, firstServerMsg.Salt, firstServerMsg.Iteration, clientNonce);
-            await scramFinalClientMsg.Write(WriteBuffer, async);
-            await WriteBuffer.Flush(async);
+            var saltBytes = Convert.FromBase64String(firstServerMsg.Salt);
+            var saltedPassword = Hi(passwd.Normalize(NormalizationForm.FormKC), saltBytes, firstServerMsg.Iteration);
+
+            var clientKey = HMAC(saltedPassword, ClientKey);
+            var storedKey = SHA256.Create().ComputeHash(clientKey);
+
+            var clientFirstMessageBare = "n=*,r=" + clientNonce;
+            var serverFirstMessage = $"r={firstServerMsg.Nonce},s={firstServerMsg.Salt},i={firstServerMsg.Iteration}";
+            var clientFinalMessageWithoutProof = "c=biws,r=" + firstServerMsg.Nonce;
+
+            var authMessage = $"{clientFirstMessageBare},{serverFirstMessage},{clientFinalMessageWithoutProof}";
+
+            var clientSignature = HMAC(storedKey, authMessage);
+            var clientProofBytes = XOR(clientKey, clientSignature);
+            var clientProof = Convert.ToBase64String(clientProofBytes);
+
+            var serverKey = HMAC(saltedPassword, ServerKey);
+            var serverSignature = HMAC(serverKey, authMessage);
+
+            var messageStr = $"{clientFinalMessageWithoutProof},p={clientProof}";
```

Comment:
It looks like everything getting interpolated is already a string, but given our recent headaches in EF Core, could you confirm that nothing here needs wrapped in the invariant culture?

Reply:
I don't think there's any issue with interpolating *strings* into strings, only other types that get formatted (numbers, dates)... I think this code should be fine - but let me know if you think otherwise.

Reply:
Nope, that's what I meant, just wanted to confirm. (Reviewed this from my phone, so no intellisence :))